### PR TITLE
Add ability to search by template with params

### PR DESCRIPTION
**Pull request by KovalM**
* [ ] Read PR [documentation](http://ostis-dev.github.io/sc-machine/dev/pr/)
* [ ] Update changelog
* [ ] Update documentation

Merge after #324

### CODE REVIEW

**Comment by deniskoronchik** [comment]
Link PR with an issue

File: sc-memory/sc-memory/sc_template_build.cpp [hunk 1]
```diff
@@ -329,13 +333,30 @@ class ScTemplateBuilder
...
+        std::string const idtf = m_context.HelperGetSystemIdtf(replacedAddr);
+        return ObjectInfo(replacedAddr, type, idtf);
+      }
+    }
+    return *templateItem;
+  }
 };
 
-ScTemplate::Result ScTemplate::FromScTemplate(ScMemoryContext & ctx, ScAddr const & scTemplateAddr)
+ScTemplate::Result ScTemplate::FromScTemplate(ScMemoryContext & ctx, ScAddr const & scTemplateAddr, const ScTemplateParams & params)
```
[the remark this answers is not in the document]
Comment (deniskoronchik) [reply sc-memory/sc-memory/sc_template_build.cpp:358]: Why we put parameters into building template from SCs? This function does one simple operation - translates scs-text as is into ScTemplate

File: sc-memory/sc-memory/sc_template_build.cpp [hunk 2]
```diff
@@ -144,9 +144,13 @@ class ScTemplateBuilder
   friend class ScTemplate;
 
 protected:
-  ScTemplateBuilder(ScAddr const & inScTemplateAddr, ScMemoryContext & inCtx)
+  ScTemplateBuilder(
```
Comment (deniskoronchik) [inline sc-memory/sc-memory/sc_template_build.cpp:147]: Search algorithm should use parameters replacement. We should not change template during build. We should have one template, but putting different parameters in search we will get different results

**Review by deniskoronchik — CHANGES_REQUESTED** [review]
I don't like implementation concept. Better to make replacements in search algorithm, but not during template generation

**Comment by KovalM** [comment]
> Link PR with an issue

Do you mean some specific created issue? If no I will create

File: CMakeLists.txt [hunk 3]
```diff
@@ -37,6 +37,8 @@ foreach( OUTPUTCONFIG ${CMAKE_CONFIGURATION_TYPES} )
 endforeach( OUTPUTCONFIG CMAKE_CONFIGURATION_TYPES )
 
 set(SC_MEMORY_SRC "${SC_MACHINE_ROOT}/sc-memory")
+set(SC_KPM_SRC "${SC_MACHINE_ROOT}/sc-kpm")
+set(SC_TOOLS_SRC "${SC_MACHINE_ROOT}/tools")
```
[the remark this answers is not in the document]
Comment (KovalM) [reply CMakeLists.txt:41]: The same reason why we put SC_MEMORY_SRC in a root cmake - we need use it in include_directories() for include it in other libraries. 

But I am not expert in cmake, probably we can find another way

File: sc-memory/sc-memory/sc_template.hpp [hunk 4]
```diff
@@ -216,23 +216,24 @@ enum class ScTemplateResultCode : uint8_t
...
 
 public:
-  ScTemplateGenParams(ScTemplateGenParams const & other) = delete;
-  ScTemplateGenParams & operator = (ScTemplateGenParams const & other) = delete;
+  ScTemplateParams & operator = (ScTemplateParams const & other) = delete;
 
-  ScTemplateGenParams() = default;
+  explicit ScTemplateParams()
+  {
+  }
```
[the remark this answers is not in the document]
Comment (KovalM) [reply sc-memory/sc-memory/sc_template.hpp]: Suggestion applied with replace ScTemplateGenParams by ScTemplateParams

File: sc-memory/sc-memory/sc_template_build.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (KovalM) [reply sc-memory/sc-memory/sc_template_build.cpp:358]: Suggestion with const modifier is applied.
 
Operation of this method is to builds template based on template in sc-memory. Probably you mixed up it with FromScs(...) method.

File: sc-memory/sc-memory/sc_template.hpp [hunk 5]
```diff
@@ -251,7 +252,7 @@ class ScTemplateGenParams
     return false;
   }
 
-  SC_DEPRECATED(0.4.0, "You should to use ScTemplateGenParams::IsEmpty")
+  SC_DEPRECATED(0.4.0, "You should to use ScTemplateParams::IsEmpty")
```
[the remark this answers is not in the document]
Comment (KovalM) [reply sc-memory/sc-memory/sc_template.hpp]: Removed

File: sc-memory/sc-memory/sc_template.hpp [hunk 6]
```diff
@@ -216,23 +216,24 @@ enum class ScTemplateResultCode : uint8_t
...
+  {
+  }
 
-  SC_DEPRECATED(0.4.0, "You should to use ScTemplateGenParams::Add")
-  _SC_EXTERN ScTemplateGenParams & add(std::string const & varIdtf, ScAddr const & value)
+  SC_DEPRECATED(0.4.0, "You should to use ScTemplateParams::Add")
+  _SC_EXTERN ScTemplateParams & add(std::string const & varIdtf, ScAddr const & value)
   {
     return Add(varIdtf, value);
   }
```
[the remark this answers is not in the document]
Comment (KovalM) [reply sc-memory/sc-memory/sc_template.hpp]: Removed

File: sc-memory/sc-memory/sc_template_build.cpp (hunk 2 above)
Comment (KovalM) [reply sc-memory/sc-memory/sc_template_build.cpp:147]: I agree and I tried to do it at first but I did not succeed. I will clarify what the difficulties were and we could discuss it

**Comment by KovalM** [comment]
After discussion we decided to redesign the ScTemplate concept. This pull request is not actual

**Pull request closed without merging**